packages/flutter_tools/lib/src/build_system/build_system.dart

```diff
@@ -204,7 +208,7 @@ abstract class Target {
         canSkip = false;
       }
     }
-    return canSkip;
+    return _ChangeResult(previousOutputs, canSkip);
```

Comment:
Consider caching the decoded stamp file json, and adding getters to the Target whose values are cached after being computed the first time. So, in this case, instead of returning _ChangeResult(...), it would go back to returning a bool, but Target would have a separate getter `previousOutputs` that is computed once using the decoded json that may or may not have already been cached itself.

Reply:
I think that to be ergonomic this needs some refactoring. The Target should not be used as the runtime representation of the build graph, that should be delegated to another interface which is mutable, allowing my to cache these values. That would potentially simplify the requirements for Source vs files too - Target could allow configuring sources, while the runtime representation could use Files.

Reply:
Updated this to have a runtime component which caches the outputs/inputs/stamp and deals only with files.

As a consequence, I'll need to deprecate/remove all the functionality that allowed for completely dynamic outputs (like output wildcards), but the tradeoff is worth it - I can now detect that a rule has changed only when the outputs change.

This ended up blocking the build assets with flutter assemble PR, since changing the output directory would not in of itself trigger a rebuild. After this PR it does.